packages/protocol/contracts/signal/SignalService.sol

```diff
@@ -170,6 +173,17 @@ contract SignalService is AuthorizableContract, ISignalService {
         return false;
     }
 
+    /// @notice Translate a RLP-encoded list of RLP-encoded TrieNodes into a list of LP-encoded
+    /// TrieNodes.
+    function _transcode(bytes memory proof) internal pure returns (bytes[] memory proofs) {
+        RLPReader.RLPItem[] memory nodes = RLPReader.readList(proof);
+        proofs = new bytes[](nodes.length);
+
+        for (uint256 i; i < nodes.length; ++i) {
+            proofs[i] = RLPReader.readBytes(nodes[i]);
+        }
+    }
```

Comment:
Obviously prefer not having this on-chain (and i know Daniel already opened a PR ).

We can tho manage to upgrade in a ways that: X messageId is old version implementation verification - after X this new ?
So basically having them living both side by side for some time.. Tehn upgrade the contracts again and exclude the old.

This way we can be compatible by not causing too much damage.

Reply:
We can do it in two PRs, after merging this one,  changing this PR to support the tokenID based handling  --https://github.com/taikoxyz/taiko-mono/pull/15609. The old code can still be deleted.

Reply:
Wait, the `messageId` is only visble on the Bridge level, not on the SignalService level, so we have to figure out another way ...

Reply:
I'll simply ask Korbi to use eth_getstorageat to get the SignalService's implementation address, then if address == old_impl, then use the relayer's returned proof as-is, if the addrss == new_impl, the bridge dapp should RLP decode then proof, then ABI-encode the decoded list, then use the generated data.